test.py

```diff
@@ -216,43 +225,46 @@ def build_test_list(self) -> List[str]:
         return [os.path.splitext(t.relative_to(self.suite_path))[0] for t in
                 self.suite_path.glob(self.pattern)]
 
+    async def _test_defs(self, test_list: List[str]) -> List[TestCase]:
+        """For the tests of the suite, build a list (test, case)"""
+        # If not implemented in subclass, just return the list of tests
+        return [TestCase(self.name, test_name, None) for test_name in test_list]
+
     async def add_test_list(self) -> None:
         options = self.options
-        lst = self.build_test_list()
-        if lst:
+        test_defs = await self._test_defs(self.build_test_list())
+
+        if test_defs and self.run_first_tests:
             # Some tests are long and are better to be started earlier,
             # so pop them up while sorting the list
-            lst.sort(key=lambda x: (x not in self.run_first_tests, x))
+            test_defs.sort(key=lambda x: (str(x) not in self.run_first_tests, str(x)))
 
-        pending = set()
-        for test_name in lst:
-            if test_name in self.disabled_tests:
-                continue
+        def is_enabled(name: str):
+            return not options.names or any (n in name for n in options.names)
 
-            t = os.path.join(self.name, test_name)
-            patterns = options.names if options.names else [t]
-            if options.skip_pattern and options.skip_pattern in t:
-                continue
+        def is_disabled(name: str):
+            return name in self.disabled_tests
 
-            async def add_test(test_name) -> None:
-                # Add variants of the same test sequentially
-                # so that case cache has a chance to populate
-                for i in range(options.repeat):
-                    await self.add_test(test_name)
-                    self.pending_test_count += 1
-
-            for p in patterns:
-                if p in t:
-                    pending.add(asyncio.create_task(add_test(test_name)))
-        if len(pending) == 0:
-            return
-        try:
-            await asyncio.gather(*pending)
-        except asyncio.CancelledError:
-            for task in pending:
-                task.cancel()
-            await asyncio.gather(*pending, return_exceptions=True)
-            raise
+        def should_skip(name: str):
+            return options.skip_pattern and options.skip_pattern in name
+
+        add_test_tasks = []
```

Comment:
This commit is mixing logical changes with stylistic changes including indentation, making it much harder to review than it could have been.

For example, I see that `add_test_tasks` is basically the same thing as previously `pending`.

And instead of
```
if len(pending) == 0:
    return
... logic ...
```

you have:
```
if add_test_tasks:
    ... logic ...
```

where `... logic ...` is exactly the same, just indented differently.

Such stylistic changes should really have their own commits.